projects/portal/src/app/models/cosmos/tx-common.service.ts

```diff
@@ -1,14 +1,158 @@
+    gas: proto.cosmos.base.v1beta1.ICoin | null | undefined,
+    fee: proto.cosmos.base.v1beta1.ICoin | null | undefined,
+  ): Promise<cosmosclient.TxBuilder> {
+    const sdk = await this.cosmosSDK.sdk().then((sdk) => sdk.rest);
+    const packedAnyMessages: proto.google.protobuf.IAny[] = messages.map((message) =>
+      cosmosclient.codec.packAny(message),
+    );
+    const txBody = new proto.cosmos.tx.v1beta1.TxBody({
+      messages: packedAnyMessages,
+    });
+    const authInfo = new proto.cosmos.tx.v1beta1.AuthInfo({
+      signer_infos: [
+        {
+          public_key: cosmosclient.codec.packAny(cosmosPublicKey),
+          mode_info: {
+            single: {
+              mode: proto.cosmos.tx.signing.v1beta1.SignMode.SIGN_MODE_DIRECT,
+            },
+          },
+          sequence: baseAccount.sequence,
+        },
+      ],
+      fee: {
+        amount: fee ? [fee] : null,
+        gas_limit: cosmosclient.Long.fromString(gas?.amount ? gas.amount : '200000'),
+      },
+    });
+    const txBuilder = new cosmosclient.TxBuilder(sdk, txBody, authInfo);
+    return txBuilder;
+  }
+
+  async signTx(
+    txBuilder: cosmosclient.TxBuilder,
+    signerBaseAccount: proto.cosmos.auth.v1beta1.BaseAccount,
+    currentCosmosWallet: CosmosWallet,
+    privateKey?: string,
+  ): Promise<cosmosclient.TxBuilder | undefined> {
+    if (currentCosmosWallet.type === WalletType.ununifi) {
+      return await this.signTxWithPrivateKey(txBuilder, signerBaseAccount, privateKey);
+    }
+    if (currentCosmosWallet.type === WalletType.keplr) {
+      return this.signTxWithKeplr(txBuilder, signerBaseAccount);
+    }
+    if (currentCosmosWallet.type === WalletType.ledger) {
+      return this.signTxWithLedger(txBuilder, signerBaseAccount);
+    }
+    if (currentCosmosWallet.type === WalletType.keyStation) {
+      return this.signTxWithKeyStation(txBuilder, signerBaseAccount);
+    }
+    throw Error('Unsupported wallet type!');
+  }
+
+  async signTxWithPrivateKey(
+    txBuilder: cosmosclient.TxBuilder,
+    signerBaseAccount: proto.cosmos.auth.v1beta1.BaseAccount,
+    privateKey?: string,
+  ): Promise<cosmosclient.TxBuilder | undefined> {
+    let cosmosPrivateKey: cosmosclient.PrivKey | undefined;
+    if (privateKey) {
+      cosmosPrivateKey = createCosmosPrivateKeyFromString(KeyType.secp256k1, privateKey);
+    } else {
+      const privateWallet: StoredWallet & { privateKey: string } =
+        await this.walletApplicationService.openUnunifiKeyFormDialog();
+      if (!privateWallet || !privateWallet.privateKey) {
+        this.snackBar.open('Failed to get Wallet info from dialog! Tray again!', 'Close');
+        return;
+      }
+      if (!validatePrivateStoredWallet(privateWallet)) {
+        this.snackBar.open('Invalid Wallet info!', 'Close');
+        return;
+      }
+      cosmosPrivateKey = createCosmosPrivateKeyFromString(
+        KeyType.secp256k1,
+        privateWallet.privateKey,
+      );
+      if (!privateKey) {
```

Comment:
This `if` must be removed. It will always throw an error here.

Reply:
you are right. thanks.